core/verify_manager.go

```diff
@@ -0,0 +1,150 @@
+package core
+
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/log"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+const verifiedCacheSize = 256
+
+type VerifyManager struct {
+	bc                   *BlockChain
+	tasks                map[common.Hash]*VerifyTask
+	peers                VerifyPeers
+	verifiedCache        *lru.Cache
+	allowUntrustedVerify bool
+	verifyCh             chan common.Hash
+	exitCh               chan struct{}
+}
+
+func NewVerifyManager(blockchain *BlockChain) *VerifyManager {
+	verifiedCache, _ := lru.New(verifiedCacheSize)
+	vm := &VerifyManager{
+		bc:            blockchain,
+		tasks:         make(map[common.Hash]*VerifyTask),
+		verifiedCache: verifiedCache,
+		verifyCh:      make(chan common.Hash),
+		exitCh:        make(chan struct{}),
+	}
+	return vm
+}
+
+func (vm *VerifyManager) Start() {
+	//read disk store to initial verified cache
+	//load unverified blocks in a normalized chain and start a batch of verify task
+	header := vm.bc.CurrentHeader()
+	go vm.mainLoop(header)
+}
+
+func (vm *VerifyManager) Stop() {
+	//stop all the tasks
+	close(vm.exitCh)
+}
+
+func (vm *VerifyManager) mainLoop(header *types.Header) {
+	//Start verify task from H to H-11 if need.
+	vm.NewBlockVerifyTask(header)
+	prune := time.NewTicker(time.Second)
+	defer prune.Stop()
+	for {
+		select {
+		case hash := <-vm.verifyCh:
+			vm.cacheBlockVerified(hash)
+			rawdb.WriteTrustBlockHash(vm.bc.db, hash)
+			delete(vm.tasks, hash)
+		case <-prune.C:
+			for hash, task := range vm.tasks {
+				if vm.bc.CurrentHeader().Number.Uint64()-task.blockHeader.Number.Uint64() > 15 {
+					delete(vm.tasks, hash)
+					close(task.terminalCh)
+				}
+			}
+		case <-vm.exitCh:
+			return
+		}
+	}
+}
+
+func (vm *VerifyManager) NewBlockVerifyTask(header *types.Header) {
+	for i := 0; i <= 11; i++ {
+		hash := header.Hash()
+		//if verified cache record that this block has been verified, skip.
+		if _, ok := vm.verifiedCache.Get(hash); ok {
+			header = vm.bc.GetHeaderByHash(header.ParentHash)
+			continue
+		}
+		//if there already has a verify task for this block, skip.
+		if _, ok := vm.tasks[hash]; ok {
+			header = vm.bc.GetHeaderByHash(header.ParentHash)
```

Comment:
emm, too many `header = vm.bc.GetHeaderByHash(header.ParentHash)`, can we simply it?